pip/pip-446.md

```diff
@@ -0,0 +1,532 @@
+- Uses attribute (not event) for better queryability in tracing backends
+
+## Public-facing Changes
+
+### Public API
+
+**New interfaces** (`org.apache.pulsar.client.api`):
+
+```java
+public interface TraceableMessage {
+    void setTracingSpan(io.opentelemetry.api.trace.Span span);
+    io.opentelemetry.api.trace.Span getTracingSpan();
+}
+
+public interface TraceableMessageId {
+    void setTracingSpan(io.opentelemetry.api.trace.Span span);
+    io.opentelemetry.api.trace.Span getTracingSpan();
+}
+```
+
+**ClientBuilder new methods**:
+
+```java
+/**
+ * Set the OpenTelemetry instance for observability.
+ */
+ClientBuilder openTelemetry(OpenTelemetry openTelemetry);
+
+/**
+ * Enable or disable automatic tracing.
+ * When enabled, uses GlobalOpenTelemetry.get() if no OpenTelemetry instance is set.
+ * Tracing interceptors are automatically added to all producers and consumers.
+ */
+ClientBuilder enableTracing(boolean tracingEnabled);
+```
+
+**New public classes** (`org.apache.pulsar.client.impl.tracing`):
+- `OpenTelemetryProducerInterceptor`: Producer interceptor for tracing
+- `OpenTelemetryConsumerInterceptor<T>`: Consumer interceptor for tracing
+- `TracingContext`: Utility methods for span creation
+- `TracingProducerBuilder`: Helper for manual trace context injection
+
+**Modified classes**:
+- `Message` interface: Now extends `TraceableMessage` (via implementations)
+- `MessageId` interface: Now extends `TraceableMessageId` (via implementations)
+
+### Binary protocol
+
+No changes to binary protocol. Trace context is propagated via existing message properties mechanism.
+
+### Configuration
+
+**New ClientBuilder options**:
+- `openTelemetry(OpenTelemetry)`: Set OpenTelemetry instance
+- `openTelemetry(OpenTelemetry, boolean)`: Set OpenTelemetry and enable tracing
+- `enableTracing(boolean)`: Enable automatic tracing
+
+**Example configuration**:
+
+```java
+// Option 1: Explicit OpenTelemetry instance with tracing enabled
+OpenTelemetry otel = OpenTelemetrySdk.builder()
+    .setTracerProvider(tracerProvider)
+    .build();
+
+PulsarClient client = PulsarClient.builder()
+    .serviceUrl("pulsar://localhost:6650")
+    .openTelemetry(otel)
+    .enableTracing(true)
+    .build();
+
+// Option 2: Use GlobalOpenTelemetry
+PulsarClient client = PulsarClient.builder()
+    .serviceUrl("pulsar://localhost:6650")
+    .enableTracing(true)  // Uses GlobalOpenTelemetry
+    .build();
+
+// Option 3: Manual interceptor (advanced)
+Producer<String> producer = client.newProducer(Schema.STRING)
+    .topic("my-topic")
+    .intercept(new OpenTelemetryProducerInterceptor())
+    .create();
+```
+
+### CLI
+
+No CLI changes. This is a client library feature.
+
+### Metrics
+
+This PIP focuses on distributed tracing, not metrics. No new metrics are added.
+
+# Monitoring
+
+Users can monitor tracing effectiveness through their OpenTelemetry backend (e.g., Jaeger, Zipkin, Grafana Tempo):
```

Comment:
Where can users configure the backend address? I don't see any related settings. 

 Is it through environment variables?

Reply:
Yes, that's handled with OTel configuration. Usually OTel configuration is handled with environment variables, although there are programmatic ways to construct and configure the `OpenTelemetry` instance.